Python/Product/Analysis/AnalysisHashSet.cs

```diff
@@ -505,6 +564,20 @@ struct Bucket {
             public int HashCode;        // the hash code of the contained key.
         }
 
+        struct BucketSet {
+            public BucketSet(int capacity, int count = 0) {
+                Buckets = capacity > 0 ? new Bucket[capacity] : Array.Empty<Bucket>();
+                Count = count;
+                if (count > capacity) {
+                    throw new ArgumentOutOfRangeException(nameof(count));
+                }
+            }
+
+            public readonly Bucket[] Buckets;
+            public int Count;
```

Comment:
Structs shouldn't have mutable fields

Reply:
Sure they can. In this class, we always copy it (under lock) and then copy it back (under lock) - it never gets updated directly.